Source/WebKit/NetworkProcess/storage/CacheStorageDiskStore.cpp

```diff
@@ -0,0 +1,520 @@
+        *size,
+        *requestHeadersGuard,
+        *request,
+        WTFMove(options),
+        WTFMove(*referrer),
+        *responseHeadersGuard,
+        WTFMove(response),
+        WTFMove(*responseBodySize)
+    };
+}
+
+std::optional<CacheStorageRecord> CacheStorageDiskStore::readRecordFromFileData(Vector<uint8_t>& buffer, Vector<uint8_t>& blobBuffer)
+{
+    if (buffer.isEmpty())
+        return std::nullopt;
+
+    auto fileData = Span { buffer.data(), buffer.size() };
+    auto metaData = decodeRecordMetaData(fileData);
+    if (!metaData)
+        return std::nullopt;
+
+    if (metaData->cacheStorageVersion != currentVersion || metaData->timeStamp > WallTime::now())
+        return std::nullopt;
+
+    auto headerData = fileData.subspan(metaData->headerOffset, metaData->headerSize);
+    if (metaData->headerHash != computeSHA1(headerData, m_salt))
+        return std::nullopt;
+
+    auto header = decodeRecordHeader(headerData);
+    if (!header)
+        return std::nullopt;
+
+    std::optional<WebCore::DOMCacheEngine::ResponseBody> responseBody;
+    if (metaData->isBodyInline) {
+        size_t bodyOffset = metaData->headerOffset + metaData->headerSize;
+        if (bodyOffset + metaData->bodySize != fileData.size())
+            return std::nullopt;
+
+        auto bodyData = fileData.subspan(bodyOffset);
+        if (metaData->bodyHash != computeSHA1(bodyData, m_salt))
+            return std::nullopt;
+
+        responseBody = WebCore::SharedBuffer::create(bodyData.data(), bodyData.size());
+    } else {
+        if (blobBuffer.isEmpty())
+            return std::nullopt;
+
+        auto sharedBuffer = WebCore::SharedBuffer::create(blobBuffer.data(), blobBuffer.size());
+        auto bodyData = Span { sharedBuffer->data(), sharedBuffer->size() };
+        if (metaData->bodyHash != computeSHA1(bodyData, m_salt))
+            return std::nullopt;
+
+        responseBody = sharedBuffer;
+    }
+
+    if (!responseBody)
+        return std::nullopt;
+
+    CacheStorageRecordInformation info { metaData->key, header->insertionTime, 0, 0, header->responseBodySize, header->request.url(), false, { } };
+    info.updateVaryHeaders(header->request, header->response);
+    return CacheStorageRecord { info, header->requestHeadersGuard, header->request, header->options, header->referrer, header->responseHeadersGuard, header->response.crossThreadData(), header->responseBodySize, WTFMove(*responseBody) };
+}
+
+void CacheStorageDiskStore::readAllRecords(ReadAllRecordsCallback&& callback)
+{
+    auto didReadRecordFiles = [this, protectedThis = Ref { *this }, callback = WTFMove(callback)](auto&& fileDatas, auto&& blobDatas) mutable {
+        ASSERT(fileDatas.size() == blobDatas.size());
+        Vector<CacheStorageRecord> result;
+        for (size_t index = 0; index < fileDatas.size(); ++index) {
+            if (auto record = readRecordFromFileData(fileDatas[index], blobDatas[index]))
+                result.append(WTFMove(*record));
```

Comment:
Maybe add some logging if read is failing.
I wonder whether we should remove files if they are not decodable.

Reply:
The write operation could be done asynchronously I guess.
In any case, adding some RELEASE_LOG_ERROR sounds like a good idea

Reply:
Will add loggings on read failure.